### PR TITLE
Run all tests in the `nix-shell`; eliminate docker infrastructure

### LINKED ISSUE #340: Move coarse-graining tests out of docker build
In #339 we have started to be more judicious about running tests in CI.  Not all fortran PRs touch the coarse-graining code, so there is often not a need to check that those tests pass upon each push to every PR.  One might ask why we can't just remove the coarse-graining tests from docker image since we already have some native regression tests for those features.  The answer there is that the docker tests do two things that the nix tests currently do not:
1. They exercise the code in debug mode.  This is useful for catching things like index errors and the like.
2. They test to make sure the code reproduces across different domain decompositions.

If we added nix tests in these settings, we could then remove these tests from the docker build and mark these tests with the `coarse` marker, further reducing the default time required in CI.

### PR DESCRIPTION
**Pull request by spencerkclark**
This PR refactors the build infrastructure in this repo to eliminate the need for the Docker component.  All development and testing is now done in the `nix` shell.  This should be a quality of life improvement for anyone developing the fortran model, as it no longer requires maintaining checksums in two separate build environments.

In so doing it introduces the following changes:
- New `make` rules are provided for compiling the model in different modes:
  - `build` -- build executables in `repro` (our production mode) and `debug` mode.
  - `build_repro` -- build only the `repro` mode executable.
  - `build_debug` -- build only the `debug` mode executable.
- Tests are run with each of the executables available in the local `bin` directory, and are tagged with the associated compile mode.  
- An option, `check_layout_invariance`, is provided to trigger regression tests be run with a 1x2 domain decomposition instead of a 1x1 domain decomposition to check invariance to the domain decomposition layout; this is used for the all the coarse-graining regression tests and replaces the previous `test_run_reproduces_across_layouts` test that would run in the docker image.
- `debug`-mode and `repro`-mode simulations produce different answers, which is something we noticed in #364 when upgrading compiler versions as well, and so require different reference checksums.

In working on this PR, we ran the fortran model in `debug` mode in more contexts than we had previously, some of which turned up errors, which we currently work around by using `pytest.skip` (something we had implicitly already been doing before):
- #365
- #381 

Working on this PR also brought my attention to the fact that `pytest`'s `tmpdir` fixture does not automatically get cleaned up after each test; `pytest` versions older than 7.3.0 keep around directories from the last three runs of `pytest`, which fill up disk space quickly since running these tests requires creating 10's of run directories, each with their own initial conditions and input files (#380).  For the time being I manually clean up these run directories after successful tests.

Resolves #340.

### CODE REVIEW

**Comment by spencerkclark** [comment]
After this PR is reviewed I will update the required checks to:
- Minimal fortran and wrapper tests in repro mode
- Minimal fortran tests in debug mode

File: Makefile
Comment (brianhenn) [inline Makefile:63]: I guess we no longer have a use case for generating serialized data this way? Maybe worth mentioning that we've deprecated the ability in the documentation?

File: Makefile
Comment (brianhenn) [inline Makefile:12]: I've done a quick scan of the code changes and I think I'm following along, but I'd like to give a try at checking out the branch and trying the build/test process tomorrow, since I might provide more useful feedback that way.

File: Makefile
Comment (spencerkclark) [reply Makefile:63]: I believe so.  In working on #364, after discussion with @elynnwu and @oliverwm1, we decided to remove `serialbox` from the build / test system since it was no longer needed and would require changes to run in a new base image (as an aside, the deadline we thought we had that triggered the need for that PR seems to have passed without any ill effects).  

In principle, we could keep the docker image around and only use it as a regression test of the serialization capability (say on merges to main, not on each commit to each PR).  That would cordon off the docker image in a place that we would generally not need to interact with it during everyday development, which is the main goal of this PR.  Admittedly, it would probably be good to get @oelbert's opinion on this (happy to chat offline to provide more context), since he is the main person that still might be interested in this functionality (though he may be working in forks, in SHiELD, or off specific earlier commits of this repository, which may make this less relevant).

I agree, whatever we decide, it should be made clear in the top-level README.

File: README.md
Comment (brianhenn) [inline README.md:39]: Minor: repeated text about using cachix

File: tests/pytest/test_regression.py
Comment (brianhenn) [inline tests/pytest/test_regression.py:60]: Definitely much easier to follow the tests now!

**Review by brianhenn — COMMENTED** [review]
Thanks @spencerkclark things worked well for me. Happy to approve if you like or can wait for another set of eyes.

File: tests/pytest/_regtest_outputs/test_regression.test_regression[debug-default.yml-False].out
Comment (brianhenn) [reply tests/pytest/_regtest_outputs/test_regression.test_regression[debug-default.yml-False].out:1]: This was resolved by making sure I updated the `configure.fv3.nix` file.

File: Makefile
Comment (spencerkclark) [reply Makefile:63]: I spoke with @oelbert about this offline.  He still uses the serialize functionality in the docker image on occasion, but is OK with us removing it, since it is fine for him to refer to older commits of the repository (it is not important for him to keep up with fortran changes that we make).

I will add a note of this deprecation to the README, and point to the last valid commit.

File: README.md
Comment (spencerkclark) [reply README.md:39]: Oops thanks, good catch,

**Review by brianhenn — APPROVED** [review]
Thanks @spencerkclark

**Pull request merged**
